### PR TITLE
#1 Create `SequencedExecutor`.

### LINKED ISSUE #1: SequencedExecutor
# Use cases
- Limit the number of parallel tasks of a similar nature that are executed at the same time to 1.
- Ensure mutually exclusive "lock" over a mutable resource **without blocking**
  - Done by configuring this _Executor_ with only 1 parallel task.
  - Example: Protect a mutable state inside of a service

# Why not ...

## ... use `Executors.newSingleThreadExecutor();`
Because it creates a new thread. This proposition is an _Executor_ that limits without creating any new thread. The thread pool already exist and we want to use it. In the end, the tasks are dispatched to an existing _Executor_ (which does not have to be a pool, but is the typical use).

# Specifications
- Based on an existing executor (composite)

### PR DESCRIPTION
**Pull request by Adeynack**
closes #1 , #8

### CODE REVIEW

**Comment by Adeynack** [comment]
Just realised I am missing some tests for when a task fails. I will be adding them. Please in the meantime, feel free to code-review what is already available.

File: src/test/java/com/github/adeynack/executors/SequencedExecutorTest.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+  @Test
+  public void twoTasksAreNotExecutedAtTheSameTimeWhenSubmittedFromAnExecutedTask() throws Exception {
+    IntStream.range(0, 12)
+             .forEach(i -> CompletableFuture.runAsync(createTask(i, true), sequencedExecutor));
+    startedTaskWait.get(1, TimeUnit.MINUTES);
+    Assert.assertEquals("No lock should be refused.", 0, refusedLocks.get());
+  }
+
+}
```
Comment (Adeynack) [inline src/test/java/com/github/adeynack/executors/SequencedExecutorTest.java]: Missing tests for task failure.

Something like:
- a task throwing an exception should resolve into a failed future and not make the Executor crash

**Comment by semonte** [comment]
Looks great! I don't have any comments for the code :)

When working with threads I have found the annotations like `@GuardedBy` useful. For example, [Google Error Prone](https://github.com/google/error-prone/) checks that all `@GuardedBy` annotated fields are protected by the specified lock. If not, you get an error (at compile time!) such as:

```
error: [GuardedBy] This access should be guarded by 'this.mainLock', which is not currently held
                taskQueue.add(command);
```

In the above error I had removed the surrounding `mainLock`:

```java
    @Override
    public void execute(@NotNull Runnable command) {
        if (taskSubmitted) {
            taskQueue.add(command);
        } else {
            taskSubmitted = true;
            baseExecutor.execute(createTaskWrapper(command));
        }
    }
```

My two cents!

**Comment by Adeynack** [comment]
@semonte : VERY GOOD SUGGESTION! Already made me realise a couple of errors.

**Comment by Adeynack** [comment]
Added some tests. Please @semonte , re-check. Also added `error-prone` to the build!

**Comment by Adeynack** [comment]
Wow @semonte , look at [this commit](https://github.com/Adeynack/more-executors/pull/7/commits/5719455a638e6cccf344ee1944b393c15f7d2d0d). `error-prone` found that! It is quite impressive that it understood that I was expecting an exception in my test in this kind of code. I am impressed!

**Comment by meersdavy** [comment]
Isn't it possible to delegate to a [single-thread executor service](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Executors.html#newSingleThreadExecutor--) instead of maintaining your own queue?

Something like (untested code)
```java
@Override
  public void execute(@NotNull Runnable command) {
    mainLock.lock();
    try {
      singleThreadExecutor.execute(() -> baseExecutor.execute(command));
    } finally {
      mainLock.unlock();
    }
  }
```

**Comment by Adeynack** [comment]
@meersdavy : That would not solve the problem I am solving with this `SequencedExecutor`. The answer to this question is already in the issue (#1).

**Comment by semonte** [comment]
@Adeynack we should write our own error prone for Kotlin ;)

**Pull request merged**